packages/story-editor/src/components/panels/design/videoSegment/videoSegment.js

```diff
@@ -0,0 +1,204 @@
+ */
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import { __ } from '@googleforcreators/i18n';
+import styled from 'styled-components';
+import {
+  Text,
+  Button,
+  THEME_CONSTANTS,
+  BUTTON_SIZES,
+  BUTTON_TYPES,
+  BUTTON_VARIANTS,
+  useSnackbar,
+  Slider,
+} from '@googleforcreators/design-system';
+import { useCallback, useState } from '@googleforcreators/react';
+import { createPage } from '@googleforcreators/elements';
+import { useFeature } from 'flagged';
+
+/**
+ * Internal dependencies
+ */
+import { Row as DefaultRow } from '../../../form';
+import { SimplePanel } from '../../panel';
+import { getCommonValue } from '../../shared';
+import { useLocalMedia } from '../../../../app/media';
+import { useStory } from '../../../../app/story';
+import useInsertElement from '../../../canvas/useInsertElement';
+
+const Row = styled(DefaultRow)`
+  margin-top: 2px;
+`;
+
+const StyledButton = styled(Button)`
+  padding: 12px 8px;
+`;
+
+const StyledSlider = styled(Slider)`
+  width: 100%;
+`;
+
+const HelperText = styled(Text).attrs({
+  size: THEME_CONSTANTS.TYPOGRAPHY.PRESET_SIZES.X_SMALL,
+})`
+  color: ${({ theme }) => theme.colors.fg.tertiary};
+`;
+
+function VideoSegmentPanel({ pushUpdate, selectedElements }) {
+  const resource = getCommonValue(selectedElements, 'resource');
+  const elementId = getCommonValue(selectedElements, 'id');
+  const enableSegmentVideo = useFeature('segmentVideo');
+
+  const onChangeSegmentTime = (value) => {
+    const newSegmentTime = value;
+    pushUpdate({ segmentTime: newSegmentTime }, true);
+  };
+
+  const segmentTime = getCommonValue(selectedElements, 'segmentTime', 20);
+  const insertElement = useInsertElement();
+
+  const { pages, currentPage, addPageAt, deleteElementById } = useStory(
+    ({
+      state: { pages, currentPage },
+      actions: { addPageAt, deleteElementById },
+    }) => ({
+      pages,
+      currentPage,
+      addPageAt,
+      deleteElementById,
+    })
+  );
+
+  const { segmentVideo } = useLocalMedia(({ actions: { segmentVideo } }) => ({
+    segmentVideo,
+  }));
+
+  const [isSegmenting, setIsSegmenting] = useState();
+  const showSnackbar = useSnackbar(({ showSnackbar }) => showSnackbar);
+
+  const segmentButtonText = isSegmenting
+    ? __('Segmenting…', 'web-stories')
+    : __('Segment', 'web-stories');
+
+  const handleSegmentation = useCallback(async () => {
+    let initialized = false;
+    let processedFiles = 0;
+    setIsSegmenting(true);
+    showSnackbar({
+      message: __('Video segmentation in progress', 'web-stories'),
+      dismissible: true,
+    });
+
+    const originalElementId = elementId;
+    const pageIds = pages.map(({ id }) => id);
+    const originalPageIndex = pageIds.indexOf(currentPage.id);
+
+    await segmentVideo({ resource, segmentTime }, (segmentedFiles) => {
+      // return the "addElement" function
+      // called via "onUploadSuccess" of the segmented file
+      return (element) => {
```

Comment:
This gets us "early access" to the segmented files `before` they get uploaded.  

The function passed get called here https://github.com/GoogleForCreators/web-stories-wp/blob/addcf91b3ca815534c89c0dc1707ad268065ccab/packages/story-editor/src/app/media/utils/useProcessMedia.js#L588

It gets called with the segmentedFiles  as a param so which we need to check the length in addElement to see if all the uploads are done.

Calling that function returns the " addElement function" which gets called per upload one at a time --- and we check the  segmentedFiles length.

Reply:
The `await ` is there so we can call  
```
showSnackbar({
      message: __('Inserting video segments', 'web-stories'),
      dismissible: false,
    });
    
 ```
 
 After the ` segmentVideo` returns ...

Reply:
>  It would be good to refactor that to make it more understandable.

+1 I find this code hard to follow. I am not sure of the benefit of doing it this way. I am personally not a massive fan of callback / return functions pattern in javascript. With `await` I don't think it is really needed anymore.

Reply:
Okay to hash out some ideas 

We call ffMpeg segment --- which gives us back an array of files 

At this point we have the files in the correct order
```
[001,002,003]
```

We can call 

```
await uploadMedia(segmentedFiles, { onUploadSuccess }); // I started by using this
```

Which works well for the upload 

Once those are all done 

We can do the page insert(s) ...  we know how many pages / files to insert

*This is where I think it gets slightly tricky as we need to get the uploaded assets into the correct pages*

What's the best way to do that?  --- The resource names files or maybe it was alt text have the input001,input002, input003 as names so might be able to do something with that ?  Or hopefully there's another way I haven't thought of

Reply:
Yeah that's a great question on how to do that. First time we really have to care about the upload/insertion order.

Yes, the `onUploadSuccess` is called for each file individually. And all it receives is the final `Resource` object.

I wouldn't wanna rely on individual properties of the resources (like alt text) to determine the order, it's fragile and can change later without realizing that it breaks something.

Off the top of my head, I think we can do something like this to ensure correct order:

1. In `useUploadMedia.js`, update the `.map()` callback in `uploadMedia()` so that we pass the `file`'s index in the array to `additionalData`
  Something like `additionalData.batchPosition = files.length - 1 - index;` (because the `files` array is actually reversed)
2. Then, in the same `useUploadMedia.js` file, where the `onUploadSuccess()` callback is called, pass `batchPosition` to it
3. This way, we know at which position to insert the video
4. In `onUploadSuccess` callback, insert uploaded video at `currentPageIndex + 1 + batchPosition`

I hope that works :-)

Reply:
There is currently no error handling. Error handling is important. If you have say 5 videos to upload in one go, then the likelihood one or more files will fail in higher. This would result in some weird side effects, like missing middle pages etc. For this prototype to be merged, I feel like we need to solve these problems. 

Regarding knowing the order, how about adding to the additionalData then saving this as part of the post meta. From there we can work on the order, maybe with a new attribute on the resource object.

Reply:
It is also worth noting that, file name is a good way to get position. If for example, file001.mp4 already existing on upload, is can result in a file name like this, file001-1.mp4.

Reply:
Pushed up a WIP - commit

https://github.com/GoogleForCreators/web-stories-wp/pull/12335/commits/f4b625b794a12c6c71989b91505ac5402c1ffcbd

1) added the index to the  additionalData 
2) added batchCount to the  additionalData
3) pushing those into a segmentedFiles array on upload success --- 

```
 segmentedFiles.push({ batchPosition, newResource });
```

For each uploadSuccess callback I'm checking the segmentedFiles.length against the  additionalData.batchCount  length to see if we're done

4) If we're done I'm sorting the segmentedFiles based on the batchPosition to get the array into the original order

That's all working

======

Next steps --- error handling + inserting the pages and elements

Reply:
> There is currently no error handling. [..] For this prototype to be merged, I feel like we need to solve these problems.

I'd disagree with this, since proper error reoslution very much depends on UX feedback.

> It is also worth noting that, file name is a good way to get position.

Let's not rely on things like alt text or file names. It's fragile.

> For each uploadSuccess callback I'm checking the segmentedFiles.length against the additionalData.batchCount length to see if we're done

Let's find a way to avoid mixing the logic for an individual item's success & the success of the whole batch operation.

Off the top of my head:

1. `uploadMedia()` could do something like `const batchId = uuidv4(); additionalData.batchId = batchId;` and then _return_ the `batchId` to the caller, so that we can store that in state.
2. Add a new `isBatchUploading(batchId)` selector to `useMediaUploadQueue`
3. In `videoSegment`, add a new effect that checks `isBatchUploading` and then creates the pages & inserts elements
4. For that to work, `segmentedFiles` can be in state

WDYT?

Reply:
Sounds good -- will pick this back up next week and move in the isBatchUploading direction

Reply:
Had an issue pushing to this branch ... will figure that out later

In the meantime I started moving things to an isBatchUploading check 

The queue seems to always be empty any idea why?

https://github.com/GoogleForCreators/web-stories-wp/pull/12370/files#diff-cf3c49123af8e30828deaed0e5d68b9c0c4010945833512d58aa63e85c0e0161R876

Reply:
@swissspidy 

2 questions

1) Thoughts on best way to get the batchId back to videoSegment ... to check using the useEffect
2) I added the isBatchUploading --- the queue is always empty so I must not be using it correctly or something any ideas?

Reply:
Resolving this as the code has been re-factored quite a bit now 

Picking up here

https://github.com/GoogleForCreators/web-stories-wp/pull/12335#issuecomment-1260832981